server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java

```diff
@@ -798,4 +749,114 @@ public void testUuidCreationLogging() {
             )
         );
     }
+
+    /**
+     * Tests writing multiple blobs to ShardBlobToDelete when it has a variable sized stream.
+     * Initially, if there is capacity, we write N blobs to ShardBlobToDelete. We expect each of them to be compressed
+     *  and written to the underlying stream.
+     * Once capacity is reached, we write M subsequent blobs, but expect that they will not be written to the
+     *  underlying stream.
+     * When we read from the stream, we expect only the successful writes to be returned
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.repositories.blobstore:WARN")
+    public void testWriteToShardBlobToDelete() {
+        int heapMemory = randomIntBetween(0, 20000);
+        int leakedBlobCount = 0;
+
+        client().admin()
+            .cluster()
+            .prepareUpdateSettings(TEST_REQUEST_TIMEOUT, TEST_REQUEST_TIMEOUT)
+            .setPersistentSettings(
+                Settings.builder().put("repositories.blobstore.max_heap_size_for_snapshot_deletion", heapMemory + "b").build()
+            )
+            .get();
+
+        final var repo = setupRepo();
+        try (var shardBlobsToDelete = repo.new ShardBlobsToDelete()) {
+            try (var mockLog = MockLog.capture(BlobStoreRepository.class)) {
+                final var expectedShardGenerations = ShardGenerations.builder();
+                final var expectedBlobsToDelete = new HashSet<String>();
+                CountDownLatch countDownLatch;
+                int blobCount = 0;
+
+                // Write while there is capacity, and then bound the number of leaked blobs
+                while (leakedBlobCount < 100) {
+                    // Generate the next blob to write
+                    final var indexId = new IndexId(randomIdentifier(), randomUUID());
+                    final var shardId = between(1, 30);
+                    final var shardGeneration = new ShardGeneration(randomUUID());
+                    // Always write at least one blob, guaranteeing that the shardDeleteResults stream increases in size
```

Comment:
:+1: well spotted (in fact the stream grows anyway because we write the index ID, but that doesn't necessarily increase `leakedBlobCount`)